### PR TITLE
Add time to event and deadline and get update for them working

### PR DESCRIPTION
**Pull request by Aak242**
Add start and end time as a compulsory part of event, and a deadline date and time as well. Also add update functionality for events and deadline to be able to update the dates and times along with the description and have changes to the start date show up on the UI

### CODE REVIEW

File: src/main/java/data/Event.java [hunk 1]
```diff
@@ -30,6 +35,28 @@ public String getStartDate() {
         return startDate;
     }
 
+    /**
```
Comment (Mahesh1772) [inline src/main/java/data/Event.java:38]: Good job spotting that we need start and end time getters

File: src/main/java/data/Task.java [hunk 2]
```diff
@@ -97,6 +97,26 @@ public String getEndDate() {
         return "";
     }
 
+    /**
+     * Getter for start time of Event task.
+     * It is a dummy function meant for the Event subclass to override.
+     *
+     * @return Empty string.
+     */
+    public String getStartTime() {
```
Comment (Mahesh1772) [inline src/main/java/data/Task.java:106]: You could make these into abstract methods

File: src/main/java/data/Task.java [hunk 3]
```diff
@@ -107,6 +127,16 @@ public String getByDate() {
         return "";
     }
 
+    /**
+     * Getter for by time for a Deadline task.
+     * It is a dummy function meant for the Deadline subclass to override
+     *
+     * @return Empty string.
+     */
+    public String getByTime() {
```
Comment (Mahesh1772) [inline src/main/java/data/Task.java:136]: Change these to abstract methods

**Review by Mahesh1772 — COMMENTED** [review]
LGTM

File: src/main/java/data/Event.java [hunk 4]
```diff
@@ -61,7 +88,8 @@ public String getTaskType() {
      */
     @Override
     public String getSaveFormat () {
-        return getTaskType() + "|" + getName() + "|" + getStartDate() + "|" + getEndDate();
+        return getTaskType() + "|" + getName() + "|" + getStartDate() + "|" + getEndDate() + "|"
+                + getStartTime() + "|" + getEndTime();
```
Comment (NicholasTanYY) [inline src/main/java/data/Event.java:92]: Neat!

File: src/main/java/data/TaskManager.java [hunk 5]
```diff
@@ -90,29 +101,113 @@ public void deleteTask(LocalDate date, int taskIndex) {
...
+                    System.out.println("Enter the new deadline date and time, separated by a space:");
+                    String[] newDateAndTime = scanner.nextLine().trim().split(" ");
+                    task = new Deadline(newTaskDescription, newDateAndTime[0], newDateAndTime[1]);
+
+                    logger.log(Level.INFO, "Updating task description from " +
+                            oldDescription + " to: " + newTaskDescription);
+                    logger.log(Level.INFO, "Updating task deadline date from " + oldDeadline.getByDate() + " to: "
+                            + newDateAndTime[0]);
+                    logger.log(Level.INFO, "Updating task deadline time from " + oldDeadline.getByTime() + " to: "
+                            + newDateAndTime[1]);
```
Comment (NicholasTanYY) [inline src/main/java/data/TaskManager.java:188]: Great use of logging to keep track of the updates being done!

**Review by NicholasTanYY — COMMENTED** [review]
Great job on adding timings to events and deadlines! LGTM.

File: src/main/java/data/Deadline.java [hunk 6]
```diff
@@ -27,13 +29,26 @@ public String getByDate() {
         return byDate;
     }
 
+    /**
+     * Getter for by time of Deadline task.
+     * Overrides super dummy function.
+     *
+     * @return The String representation of the by time.
+     */
+    @Override
```
Comment (kyhjonathan) [inline src/main/java/data/Deadline.java:38]: LGTM!

File: src/main/java/data/Deadline.java [hunk 7]
```diff
@@ -47,6 +62,6 @@ public String getTaskType () {
      */
     @Override
     public String getSaveFormat () {
-        return getTaskType() + "|" + getName() + "|" + getByDate();
+        return getTaskType() + "|" + getName() + "|" + getByDate() + "|" + getByTime();
```
Comment (kyhjonathan) [inline src/main/java/data/Deadline.java:65]: Good job in adding the additional method

File: src/main/java/data/TaskManager.java [hunk 8]
```diff
@@ -90,29 +101,113 @@ public void deleteTask(LocalDate date, int taskIndex) {
...
      */
-    public static void updateTask(LocalDate date, int taskIndex, String newTaskDescription)
+    public static void updateTask(LocalDate date, int taskIndex, String newTaskDescription, Scanner scanner)
             throws IndexOutOfBoundsException {
         try {
             List<Task> dayTasks = tasks.get(date);
             boolean dayHasTasks = dayTasks != null;
             boolean taskIndexExists = taskIndex >= 0 && taskIndex < Objects.requireNonNull(dayTasks).size();
-            assert dayTasks != null;
+            assert dayHasTasks;
```
Comment (kyhjonathan) [inline src/main/java/data/TaskManager.java:113]: Good assertions made!

File: src/main/java/data/TaskManager.java [hunk 9]
```diff
@@ -90,29 +101,113 @@ public void deleteTask(LocalDate date, int taskIndex) {
...
             assert taskIndexExists;
 
             String oldDescription = dayTasks.get(taskIndex).getName();
+            String currentTaskType = dayTasks.get(taskIndex).getTaskType();
+            boolean startDateChanged = false;
+
+            Task task;
+            switch (currentTaskType) {
```
Comment (kyhjonathan) [inline src/main/java/data/TaskManager.java:121]: Neat code!

File: src/main/java/data/TaskManager.java [hunk 10]
```diff
@@ -90,29 +101,113 @@ public void deleteTask(LocalDate date, int taskIndex) {
...
+            case "D":
+                Deadline oldDeadline = (Deadline) dayTasks.get(taskIndex);
+                System.out.println("Do you want to update the deadline date and time? (yes/no)");
+                String deadlineResponse = scanner.nextLine().trim().toLowerCase();
+                if (deadlineResponse.equals("yes")) {
+                    System.out.println("Enter the new deadline date and time, separated by a space:");
+                    String[] newDateAndTime = scanner.nextLine().trim().split(" ");
+                    task = new Deadline(newTaskDescription, newDateAndTime[0], newDateAndTime[1]);
+
+                    logger.log(Level.INFO, "Updating task description from " +
```
Comment (kyhjonathan) [inline src/main/java/data/TaskManager.java:183]: Good job on the comprehensive logging!

File: src/main/java/data/TaskManager.java [hunk 11]
```diff
@@ -90,29 +101,113 @@ public void deleteTask(LocalDate date, int taskIndex) {
...
+                            oldDescription + " to: " + newTaskDescription);
+                    logger.log(Level.INFO, "Updating task start date from " +
+                            oldEvent.getStartDate() + " to: " + newDatesAndTimes[0]);
+                    logger.log(Level.INFO, "Updating task end date from " + oldEvent.getEndDate() + " to: " +
+                            newDatesAndTimes[1]);
+                    logger.log(Level.INFO, "Updating task start time from " + oldEvent.getStartTime() + " to: " +
+                            newDatesAndTimes[2]);
+                    logger.log(Level.INFO, "Updating task end time from " + oldEvent.getEndTime() + " to: " +
```
Comment (clarencepohh) [inline src/main/java/data/TaskManager.java:163]: Good use of logging to keep track of updates

**Review by clarencepohh — COMMENTED** [review]
Thanks for implementing deadline updates 
Overall looks good. Will we be implementing time formats for the new start and end times?

**Comment by NicholasTanYY** [comment]
Removing this PR, as the merge is done with PR #95 instead

**Pull request closed without merging**